_posts/2024-08-27-building-browser.md

```diff
@@ -0,0 +1,174 @@
+In this post, we’ll explain what we tried to achieve, what we found, and what’s next for building a browser using Servo as a web engine.
+
+## Multi-view
+
+Of course, the first major feature we want to achieve is multiple webviews.
+A webview is a term abstracted from the [top-level browsing context](https://doc.servo.org/servo/struct.TopLevelBrowsingContextId.html).
+This is what people refer to as a web page.
+With multi-view support, we can create multiple web pages as tabs in a single window.
+Most importantly, we can draw our UI with additional webviews.
+The main reason we want to write UI using Servo itself is that we can dogfood our own stack and verify that it can meet practical requirements, such as prompt windows, context menus, file selectors, and more.
+
+Basic multi-view support was reviewed and merged into Servo earlier this year thanks to [@delan](https://github.com/delan) ([#30840](https://github.com/servo/servo/pull/30840), [#30841](https://github.com/servo/servo/pull/30841), [#30842](https://github.com/servo/servo/pull/30842)).
+Verso refined that into a specific type called [WebView](https://docs.versotile.org/verso/webview/struct.WebView.html).
+From there, any function that owns webviews can decide how to present them depending on their IDs.
+In a Verso window, two webviews are created at the moment—one for [handling regular web pages](https://docs.versotile.org/verso/window/struct.Window.html#method.handle_servo_messages_with_webview) and the other for [handling the UI](https://docs.versotile.org/verso/window/struct.Window.html#method.handle_servo_messages_with_panel), which is currently called the Panel.
+The result of the [showcase](https://github.com/versotile-org/verso?tab=readme-ov-file#verso) in Verso’s README.md looks like this:
+
+<figure class="_fig"><a href="{{ '/img/blog/verso-crt-demo.webp' | url }}"><img src="{{ '/img/blog/verso-crt-demo.webp' | url }}"
+    alt="Verso displaying ASCII text in a CRT style"></a>
+<figcaption>Figure 1: Verso window displaying two different webviews. One for the UI, the other for the web page.</figcaption></figure>
+
+For now, the inter-process communication is done via Servo’s existing channel messages like [EmbedderMsg](https://doc.servo.org/embedder_traits/enum.EmbedderMsg.html) and [EmbedderEvent](https://doc.servo.org/compositing/windowing/enum.EmbedderEvent.html).
+We are looking to improve the IPC mechanism with more granular control over DOM elements.
+So, the panel UI can be updated based on the status of web pages.
+One example is when the page URL is changed and the navigation bar needs to be updated.
+There are some candidates for this, such as [WebDriverCommandMsg](https://doc.servo.org/script_traits/enum.WebDriverCommandMsg.html).
+[@webbeef](https://github.com/webbeef) also started a discussion about [defining custom elements](https://github.com/servo/servo/discussions/32883) like `<web-view>` for better ergonomics.
```

Comment:
```suggestion
[@webbeef](https://github.com/webbeef) also started a discussion about [defining custom elements](https://github.com/servo/servo/discussions/32883) like `<webview>` for better ergonomics.
```

This is a suggestion, but I've seen `<webview>` used in other projects: https://www.electronjs.org/docs/latest/api/webview-tag

Reply:
Added in [f0e7fa9](https://github.com/servo/servo.org/pull/171/commits/f0e7fa9024bb157c7bfd66bda46bd04d9a84522e)